ansi/halfblocks/encoder.go

```diff
@@ -0,0 +1,626 @@
+
+	// Get the coverage pattern for the selected character
+	var coverage []bool
+	for _, b := range r.Blocks {
+		if b.Char == bestChar {
+			coverage = b.Coverage
+			break
+		}
+	}
+
+	// Assign pixels to foreground or background based on the character's coverage
+	for i := 0; i < 4; i++ {
+		y, x := i/2, i%2
+		if coverage[i] {
+			fgPixels = append(fgPixels, block.Pixels[y][x])
+		} else {
+			bgPixels = append(bgPixels, block.Pixels[y][x])
+		}
+	}
+
+	// Calculate average colors
+	if len(fgPixels) > 0 {
+		block.BestFgColor = r.averageColors(fgPixels)
+	} else {
+		// Default to black if no foreground pixels
+		block.BestFgColor = color.RGBA{0, 0, 0, 255}
+	}
+
+	if len(bgPixels) > 0 {
+		block.BestBgColor = r.averageColors(bgPixels)
+	} else {
+		// Default to black if no background pixels
+		block.BestBgColor = color.RGBA{0, 0, 0, 255}
+	}
+
+	block.BestSymbol = bestChar
+}
+
+// averageColors calculates the average color from a slice of colors
+func (r *Renderer) averageColors(colors []color.RGBA) color.RGBA {
+	if len(colors) == 0 {
+		return color.RGBA{0, 0, 0, 255}
+	}
+
+	var sumR, sumG, sumB, sumA uint32
+
+	for _, c := range colors {
+		sumR += uint32(c.R)
+		sumG += uint32(c.G)
+		sumB += uint32(c.B)
+		sumA += uint32(c.A)
+	}
+
+	count := uint32(len(colors))
+	return color.RGBA{
+		R: uint8(sumR / count),
+		G: uint8(sumG / count),
+		B: uint8(sumB / count),
+		A: uint8(sumA / count),
+	}
+}
+
+// formatSymbol formats a symbol with ANSI color codes
+func (r *Renderer) formatSymbol(char rune, fg, bg color.RGBA) string {
+	if r.Options.ColorMode == 0 {
+		// No color, just return the character
+		return string(char)
+	}
+
+	var fgStr, bgStr string
+
+	switch r.Options.ColorMode {
+	case 1: // 8 colors
+		fgCode := nearestAnsi8Color(fg.R, fg.G, fg.B)
+		bgCode := nearestAnsi8Color(bg.R, bg.G, bg.B)
+		fgStr = fmt.Sprintf("\033[%dm", 30+fgCode)
+		bgStr = fmt.Sprintf("\033[%dm", 40+bgCode)
+
+	case 2: // 256 colors
+		fgCode := nearestAnsi256Color(fg.R, fg.G, fg.B)
+		bgCode := nearestAnsi256Color(bg.R, bg.G, bg.B)
+		fgStr = fmt.Sprintf("\033[38;5;%dm", fgCode)
+		bgStr = fmt.Sprintf("\033[48;5;%dm", bgCode)
+
+	case 3: // True color
+		fgStr = fmt.Sprintf("\033[38;2;%d;%d;%dm", fg.R, fg.G, fg.B)
+		bgStr = fmt.Sprintf("\033[48;2;%d;%d;%dm", bg.R, bg.G, bg.B)
+	}
+
+	return fgStr + bgStr + string(char) + "\033[0m"
+}
+
+// getPixelSafe returns the color at (x,y) or black if out of bounds
+func (r *Renderer) getPixelSafe(img image.Image, x, y int) color.RGBA {
+	bounds := img.Bounds()
+	if x < bounds.Min.X || x >= bounds.Max.X || y < bounds.Min.Y || y >= bounds.Max.Y {
+		return color.RGBA{0, 0, 0, 255}
+	}
+
+	r8, g8, b8, a8 := img.At(x, y).RGBA()
+	return color.RGBA{
+		R: uint8(r8 >> 8),
+		G: uint8(g8 >> 8),
+		B: uint8(b8 >> 8),
+		A: uint8(a8 >> 8),
+	}
+}
+
+// scaleImage resizes an image to the specified dimensions
+func (r *Renderer) scaleImage(img image.Image, width, height int) image.Image {
```

Comment:
All scale functions should adhere to [`draw.Scaler`](https://pkg.go.dev/golang.org/x/image/draw#Scaler)